### PR TITLE
Streams

### PR DESCRIPTION
**Pull request by kartelpol**
https://epa.ms/nodejs18-hw3

### CODE REVIEW

File: src/utils/streams/errorMsgs.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+    params: {
+        wrong_format: 'INVALID ARGUMENTS COMBINATION',
+        wrong_number: (arg) => `NEED AN EXTRA ARGUMENT --${arg}. SEE --HELP (-H)`,
+        wrong_extention: (ext) => `FORMAT ERROR. EXPECTED ${ext.toUpperCase} FILE`,
+        wrong_path: 'PATH IS NOT EXIST'
```
Comment (deser) [inline src/utils/streams/errorMsgs.js]: does not

File: app.js [hunk 2]
```diff
@@ -1,9 +1,43 @@
...
+import convert from './src/utils/convertCsvToJson';
+import * as config from './config/taskInfo.json';
+
 import { Product, User } from 'models';
 
+import Dirwatcher from './src/dirwatcher';
+import Importer from './src/importer';
 
 logger.info(config.name);
+console.log(process.argv);
```
Comment (deser) [inline app.js]: logger or console.log? :)

File: config/webpack.config.js [hunk 3]
```diff
@@ -14,7 +14,8 @@ const clean = new CleanWebpackPlugin(['build'], {
     root: PROJECT_ROOT,
 });
 
-module.exports = (env, argv) => ({
+module.exports = (env, argv) => {console.log(argv); return({
```
Comment (deser) [inline config/webpack.config.js]: logger

File: src/utils/streams/actions.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
+const { Writable } = require('stream');
```
Comment (deser) [inline src/utils/streams/actions.js]: const { Writable,Transform } = require('stream');

File: src/utils/streams/actions.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+const csvToJsonConverter = require('csvtojson');
+const multistream = require('multistream');
+
+const errorMsgs = require('./errorMsgs');
+
+const BUNDLE_CSS_NAME = 'bundle.css';
+const FINAL_CSS_NAME = 'end-bundle-with-this.css';
+
+function reverse() {
```
Comment (deser) [inline src/utils/streams/actions.js]: image

Please, follow interface (at least, reverse should take string as argument)

File: src/utils/streams/actions.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        } catch(e) {
+            console.error(e);
+        }
+    } else {
+        console.error(errorMsgs.params.wrong_extention('.csv'));
+    }
+}
+
+function convertToFile(filePath) {
+    filePath = join(__dirname, filePath);
```
Comment (deser) [inline src/utils/streams/actions.js]: think how to eliminate code duplication between convertFromFile\convertToFile

File: src/utils/streams/actions.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        console.error(errorMsgs.params.wrong_extention('.csv'));
+    }
+}
+
+async function buildCss(dirPath) {
+    const relativeDirPath = relative(__dirname, dirPath);
+    const outputFile = join(__dirname, relativeDirPath, BUNDLE_CSS_NAME);
+    dirPath = join(__dirname, relativeDirPath);
+
+    const styleStreams = await getCssStreamsFromDir(dirPath);
```
Comment (deser) [inline src/utils/streams/actions.js]: try\catch

File: src/utils/streams/actions.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    styleStreams.push(getFinalCssStream());
+ 
+    const writeStream = createWriteStream(outputFile);
+
+    multistream(styleStreams).pipe(writeStream);
+}
+
+function getCssStreamsFromDir(dirPath) {
+    const styleStreams = [];
```
Comment (deser) [inline src/utils/streams/actions.js]: move it's definition inside promise

File: src/utils/streams/actions.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+            resolve(styleStreams);
+        });
+    });
+}
+
+function getFinalCssStream() {
+    const requestedCSSPath = join(__dirname, FINAL_CSS_NAME);
+    try {
+        return readStream = createReadStream(requestedCSSPath);
```
Comment (deser) [inline src/utils/streams/actions.js]:
```suggestion
        return createReadStream(requestedCSSPath);
```

File: src/utils/streams/index.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+const errorMsgs = require('./errorMsgs');
+const actions = require('./actions');
+
+analyzeArgs(args);
+
+function analyzeArgs (args) {
+    if (process.argv[2] === '--help' || process.argv[2] === '-h' || process.argv.length == 2) {
+        parser.printHelp();
+    } else if (args.action && (process.argv[2].indexOf('--action=') == 0 || process.argv[2].indexOf('-a=') == 0)) {
+        checkAction(args);
```
Comment (deser) [inline src/utils/streams/index.js]: pass only action name, not the whole `args`

**Review by deser — CHANGES_REQUESTED** [review]
Good job. Please fix comments and write tests

File: src/utils/streams/actions.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+          process.stdout.write(chunk.toString().split("").reverse().join(""));
+          callback();
+        }
+    });
+      
+    process.stdin.pipe(outStream);
+}
+
+function transform() {
+    const upperCaseTransformingStream = new Transform({
```
Comment (deser) [inline src/utils/streams/actions.js]: they say you need to use `through2`

Comment (kartelpol) [reply src/utils/streams/actions.js]: It's just an example as I understood

File: src/utils/streams/index.js (hunk 10 above)
Comment (kartelpol) [reply src/utils/streams/index.js]: Would be better to use global args inside the checkAction?

File: src/utils/streams/actions.js (hunk 5 above)
Comment (kartelpol) [reply src/utils/streams/actions.js]: I just didn't understand why I need to read stdin outside the action functions and then pass inside chunks?

**Pull request merged**